esp-hal/src/uart/uhci.rs

```diff
@@ -0,0 +1,708 @@
+    fn give_uhci(&self) -> &peripherals::UHCI0<'_> {
+        match &self.0 {
+            any::Inner::Uhci0(x) => x,
+        }
+    }
+}
+
+/// A configuration error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+pub enum ConfigError {
+    /// chunk_limit is above 4095, this is not allowed (hardware limit)
+    AboveReadLimit,
+}
+
+/// UHCI Configuration
+#[derive(Debug, Clone, Copy, procmacros::BuilderLite)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+pub struct Config {
+    /// If this is set to true UHCI will end the payload receiving process when UART has been in
+    /// idle state.
+    idle_eof: bool,
+    /// If this is set to true UHCI decoder receiving payload data ends when the receiving
+    /// byte count has reached the specified value (in len_eof).
+    /// If this is set to false UHCI decoder receiving payload data is end when 0xc0 is received.
+    len_eof: bool,
+    /// The limit of how much to read in a single read call. It cannot be higher than the dma
+    /// buffer size, otherwise uart/dma/uhci will freeze. It cannot exceed 4095 (12 bits), above
+    /// this value it will simply also split the readings
+    chunk_limit: u16,
+}
+
+impl Default for Config {
+    fn default() -> Config {
+        Config {
+            idle_eof: true,
+            len_eof: true,
+            // This is the default in the register at boot, still should be changed!
+            chunk_limit: 128,
+        }
+    }
+}
+
+impl core::error::Error for ConfigError {}
+
+impl core::fmt::Display for ConfigError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ConfigError::AboveReadLimit => {
+                write!(
+                    f,
+                    "The requested read limit is not possible. The max is 4095 (12 bits)"
+                )
+            }
+        }
+    }
+}
+
+impl<Dm> embassy_embedded_hal::SetConfig for Uhci<'_, Dm>
+where
+    Dm: DriverMode,
+{
+    type Config = Config;
+    type ConfigError = ConfigError;
+
+    fn set_config(&mut self, config: &Self::Config) -> Result<(), Self::ConfigError> {
+        self.apply_config(config)
+    }
+}
+
+/// UHCI (To use with UART over DMA)
+pub struct Uhci<'d, Dm>
+where
+    Dm: DriverMode,
+{
+    uart: Uart<'d, Dm>,
+    uhci: AnyUhci<'static>,
+    channel: Channel<Dm, PeripheralDmaChannel<AnyUhci<'d>>>,
+}
+
+impl<'d, Dm> Uhci<'d, Dm>
+where
+    Dm: DriverMode,
+{
+    fn init(&self) {
+        self.clean_turn_on();
+        self.reset();
+        self.conf_uart();
+    }
+
+    fn clean_turn_on(&self) {
+        // General conf registers
+        let reg: &uhci0::RegisterBlock = self.uhci.give_uhci().register_block();
+        reg.conf0().modify(|_, w| w.clk_en().set_bit());
+        reg.conf0().modify(|_, w| {
+            unsafe { w.bits(0) };
+            w.clk_en().set_bit()
+        });
```

Comment:
Is this supposed to be a `reg.conf0().write()`? Would `write` set an incorrect default?

Reply:
Fair enough, although I think it could still be a `write` if you overwrite all bits, there's no need to read them first.